### PR TITLE
Workspace optimizations and refactoring

### PR DESCRIPTION
**Pull request by AlexDBlack**
***COMPLETE BUT REQUIRES ADDITIONAL QA BEFORE MERGING***
All deeplearning4j-core tests are passing (CPU) but additional testing is required.

ND4J PR to go with this: https://github.com/deeplearning4j/nd4j/pull/2823

This PR overhauls workspaces in DL4J, to improve performance and memory use.
It also adds a lot of workspace validation, to detect invalid workspace use (bugs, bad layer implementations etc).

Goals here:
1. Eliminate performance bottleneck of migrateInputs() etc: all arrays are allocated in the final workspace they should be in
2. Reduce memory requirements
3. Improve maintainability and considerably simplify workspace usage:
    a. Workspace-related code had become unweildy in MultiLayerNetwork and ComputationGraph FF/BP methods, and has been simplified (and made more maintainable)
    b. A lot of additional workspace-related validation has been added - this should make it considerably easier to detect workspace-related bugs in the future (due to bugs in layers, etc)

In terms of  design:
1. Workspaces-related code should be handled at the level of array types (input, activations, etc) not workspace names ("loop_ff" etc). Layers etc don't need to know/care about the workspace configuration - simply that input arrays should use the INPUT enumeration, activation arrays should use the ACTIVATION enumeration, etc.
2. Layers, preprocessors, etc will know about workspaces. This means API changes
3. I've also slightly simplified the layer API by cleaning up a number of superfluous methods (unnecessary activate/output methods)
4. We're looking at substantial rewrite of the MLN and CG feedforward and backprop methods:
    a. They are built with workspaces in mind from scratch.
    b. Reduced code redundancy (we had essentially the same FF loops in a number of places, with minor variants)
5. Important note: I've deprecated WorkspaceMode.SINGLE/SEPARATE and added WorkspaceMode.ENABLED - single/separate modes don't really exist anymore

Also fixes:
* https://github.com/deeplearning4j/deeplearning4j/issues/4931
* https://github.com/deeplearning4j/deeplearning4j/issues/4913

### CODE REVIEW

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/eval/EvaluationUtils.java [hunk 1]
```diff
@@ -174,7 +176,7 @@ public static INDArray reshapeTimeSeriesTo2d(INDArray labels) {
             return new Pair<>(labels2d, predicted2d);
         }
 
-        INDArray oneDMask = TimeSeriesUtils.reshapeTimeSeriesMaskToVector(outputMask);
+        INDArray oneDMask = TimeSeriesUtils.reshapeTimeSeriesMaskToVector(outputMask, LayerWorkspaceMgr.noWorkspacesImmutable(), ArrayType.INPUT);  //TODO
```
[the remark this answers is not in the document]
Comment (AlexDBlack) [reply deeplearning4j-nn/src/main/java/org/deeplearning4j/eval/EvaluationUtils.java]: Fixed, just an old comment.

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/api/FwdPassType.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package org.deeplearning4j.nn.api;
+
+public enum FwdPassType {
+    STANDARD,
+    RNN_TIMESTEP,
+    RNN_ACT_STORED
```
[the remark this answers is not in the document]
Comment (AlexDBlack) [reply deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/api/FwdPassType.java]: Yep, good idea... renamed. 👍

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/workspace/ArrayType.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
+package org.deeplearning4j.nn.workspace;
+
+/**
+ * Array type enumeration for use with {@link LayerWorkspaceMgr}<br>
+ * <br>
+ * Array types:<br>
+ * INPUT: The array set to the input field of a layer (i.e., input activations)<br>
+ * ACTIVATIONS: The output activations for a layer's feed-forward method<br>
+ * ACTIVATION_GRAD
+ * FF_WORKING_MEM
```
[the remark this answers is not in the document]
Comment (AlexDBlack) [reply deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/workspace/ArrayType.java]: Fixed, good catch. I intended to fill out the rest :)

**Comment by AlexDBlack** [comment]
@maxpumperla 
> The only real criticism I have is that changing the API so that "almost everything" takes a workspace manager (WSM) seems a little excessive at first glance.

> I'm sure you had your reasons, maybe it doesn't really work any other way. My hunch says it should've been possible to set a WSM config at layer and network level.

It isn't the greatest for (developer) usability, sure - though doesn't impact end users unless they write custom layers etc as it's internal.

I initially considered a "Layer.setWorkspaceManager" type design, but eventually rejected this as likely to introduce bugs. We'd be continually swapping out workspace managers in the layers (note different feed-forward methods have very different workspace configs), and if we forget to do it at some point, we can end up with crashes and really hard to track down bugs and performance issues.
tl;dr I went with explicit and stateless as it's the least likely to introduce bugs and hidden performance/memory issues (unnecessary/unintended detaches), at the expense of those API changes.


> Another idea worth exploring is to keep the old methods and wrap them like this:

Good question, but I also considered and rejected it. I did it in one place (time series utils) but I'm not adding this to the layers, again to avoid hidden performance/memory issues. Also we have too many Layer methods already, and I'm very hesitant to add even more :)
